diff --git a/common/define-grammar.js b/common/define-grammar.js
index efb0b76..0fb6f5b 100644
--- a/common/define-grammar.js
+++ b/common/define-grammar.js
@@ -116,6 +116,7 @@ module.exports = function defineGrammar(dialect) {
         [$.primary_type, $.type_parameter],
       ] : [
         [$.jsx_opening_element, $.type_parameter],
+        [$.jsx_fragment, $.type_parameter],
         [$.jsx_namespace_name, $.primary_type],
       ],
     ),
@@ -269,6 +270,20 @@ module.exports = function defineGrammar(dialect) {
         '/>',
       )),
 
+      jsx_fragment: $ => seq(
+        field('open_tag', seq($._jsx_start_opening_element, '>')),
+        repeat($._jsx_child),
+        field('close_tag', $.jsx_closing_element)
+      ),
+
+      _jsx_child: $ => choice(
+        $.jsx_element,
+        $.jsx_self_closing_element,
+        $.jsx_text,
+        $.jsx_expression,
+        prec(1, $.mx_element)
+      ),
+
       export_specifier: (_, previous) => seq(
         optional(choice('type', 'typeof')),
         previous,
